### PR TITLE
Let Declarations inherit from RangeNode

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Make range information reliably available for all declarations, by letting the constructors of declarations take a `RangeToken` argument

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/AST/BuiltIns.cs [hunk 1]
```diff
@@ -110,8 +110,8 @@ public class BuiltIns {
     if (!ArrowTypeDecls.ContainsKey(arity)) {
       IToken tok = Token.NoToken;
```
Comment (stefan-aws) [inline Source/DafnyCore/AST/BuiltIns.cs:111]: Is this variable still needed?

File: Source/DafnyCore/AST/BuiltIns.cs [hunk 2]
```diff
@@ -152,11 +152,11 @@ public class BuiltIns {
       // declaration of total arrow-type
 
       tps = Util.Map(Enumerable.Range(0, arity + 1), x => x < arity ?
-        new TypeParameter(tok, "T" + x, TypeParameter.TPVarianceSyntax.Contravariance) :
-        new TypeParameter(tok, "R", TypeParameter.TPVarianceSyntax.Covariant_Strict));
+        new TypeParameter(RangeToken.NoToken, new Name("T" + x), TypeParameter.TPVarianceSyntax.Contravariance) :
+        new TypeParameter(RangeToken.NoToken, new Name("R"), TypeParameter.TPVarianceSyntax.Covariant_Strict));
       tys = tps.ConvertAll(tp => (Type)(new UserDefinedType(tp)));
       id = new BoundVar(tok, "f", new UserDefinedType(tok, partialArrow.Name, partialArrow, tys));
```
Comment (stefan-aws) [inline Source/DafnyCore/AST/BuiltIns.cs:158]: Could you replace `tok` here as well?

**Review by stefan-aws — APPROVED** [review]

File: Source/DafnyCore/AST/BuiltIns.cs (hunk 1 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/AST/BuiltIns.cs:111]: It's used in this method.

File: Source/DafnyCore/AST/BuiltIns.cs (hunk 2 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/AST/BuiltIns.cs:158]: I can't yet, because many of the AST nodes still take an IToken and not a RangeToken as a first argument.

**Review by stefan-aws — APPROVED** [review]

**Pull request merged**